storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3109,10 +3109,1745 @@ buf_block_t *fsp_get_latched_xdes_page(
+    {
+fetch_next_page:
+      if (next_page_no == FIL_NULL)
+        break;
+      mtr->commit();
+      cur_page_no= next_page_no;
+
+      mtr->start();
+      mtr->x_lock_space(space);
+      block= fsp_get_latched_page(page_id_t{0, cur_page_no},
+                                  mtr, &err);
+      if (!block)
+        return err;
+      continue;
+    }
+
+    uint32_t new_extent= space_defrag->get_new_extent(cur_extent);
+    /* There is no need for extent to be changed */
+    if (new_extent == cur_extent)
+    {
+      if (level)
+      {
+        err= store_child_pages(block);
+        if (err) return err;
+      }
+      goto fetch_next_page;
+    }
+
+    buf_block_t *header_block=
+      fsp_get_latched_page(page_id_t{0, 0}, mtr, &err);
+    if (!header_block)
+      return err;
+
+    const fseg_header_t *seg_header= m_root->page.frame +
+      (level ? PAGE_HEADER + PAGE_BTR_SEG_TOP
+             : PAGE_HEADER + PAGE_BTR_SEG_LEAF);
+
+    buf_block_t *iblock;
+    fseg_inode_t *inode= fseg_inode_try_get(seg_header, 0, 0, mtr,
+                                            &iblock, &err);
+    if (!inode)
+      return err;
+
+    auto parent_it= m_parent_pages.find(cur_page_no);
+    if (parent_it == m_parent_pages.end())
+    {
+      err= DB_CORRUPTION;
+      return err;
+    }
+
+    uint32_t parent_page_no= uint32_t(parent_it->second);
+
+    uint32_t parent_offset= uint32_t(parent_it->second >> 32);
+
+    if (parent_offset > srv_page_size - FIL_PAGE_DATA_END)
+    {
+      err= DB_CORRUPTION;
+      return err;
+    }
+
+    pageAllocator allocator(header_block, iblock, inode, extent_size,
+                            mtr);
+
+    associatedPages related_pages(block, mtr);
+    err= allocator.prepare(new_extent, old_state == XDES_FSEG);
+
+    DBUG_EXECUTE_IF("allocation_prepare_fail", err= DB_CORRUPTION;);
+    if (err) return err;
+
+    err= related_pages.prepare(parent_page_no);
+    DBUG_EXECUTE_IF("relation_page_prepare_fail", err= DB_CORRUPTION;);
+
+    if (err) return err;
+
+    allocator.complete();
+
+    /* After allocating the new page, try to prepare the steps
+    of page removal function. Because there is a possiblity that
+    last block in FSEG_NOT_FULL/FSP_FREE_FRAG/FSP_FREE last block
+    could've changed while allocating the new block. */
+    pageRemover remover(header_block, iblock, inode, extent_size,
+                        cur_page_no, mtr);
+    err= remover.prepare();
+ 
+    DBUG_EXECUTE_IF("remover_prepare_fail", err= DB_CORRUPTION;);
+    if (err)
+    {
+      allocator.restore_old_pages();
+      mtr->discard_modifications();
+      return err;
+    }
+
+    /* Copy the data from old block to new block */
+    buf_block_t *new_block= allocator.get_new_block();
+    uint32_t new_page_no= new_block->page.id().page_no();
+    /* Copy FIL_PAGE_PREV, FIL_PAGE_NEXT */
+    mtr->memcpy(*new_block, new_block->page.frame + FIL_PAGE_PREV,
+                block->page.frame + FIL_PAGE_PREV, 8);
```

Comment:
This is unnecessarily writing 4 extra bytes if `FIL_PAGE_NEXT` was `FIL_NULL`. Only from the start of a write we are omitting bytes that did not actually change.

I would suggest to replace the last parameter `8` with `page_has_next(block->page.frame) ? 8 : 4`.

It seems that the default template parameter `mtr_t::NORMAL` is appropriate here; if both pointers were `FIL_NULL` then this should actually be the index root page and not be subject to relocation. There is a slim chance that we were missing a call to `btr_discard_only_page_on_level()` at some point. In any case, using `mtr_t::MAYBE_NOP` would only silence a debug assertion and not make any difference in a non-debug build.